### PR TITLE
Add output log of skipped / failed components

### PR DESCRIPTION
**Pull request by cobya**
Add output log of skipped or failed components to the end of the build post detector summary. A log group will be created for each detector which has failed components.

image

### CODE REVIEW

File: src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs [hunk 1]
```diff
@@ -96,10 +114,17 @@ public sealed class SingleFileComponentRecorder : ISingleFileComponentRecorder
 
         private readonly ConcurrentDictionary<string, DetectedComponent> detectedComponentsInternal = new ConcurrentDictionary<string, DetectedComponent>();
 
+        /// <summary>
+        /// Dictionary of components which had an error during parsing and a dummy data value that only allocates 1 byte.
+        /// </summary>
+        private readonly ConcurrentDictionary<string, byte> skippedComponentsInternal = new ConcurrentDictionary<string, byte>();
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs:120]: Why not just use a `ConcurrentBag<string>`?

File: src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs [hunk 2]
```diff
@@ -173,6 +204,19 @@ public DetectedComponent GetComponent(string componentId)
...
         }
 
+        public void RegisterPackageParseFailure(string skippedComponent)
+        {
+            if (skippedComponent == null)
+            {
+                throw new ArgumentNullException(paramName: nameof(skippedComponent));
+            }
+
+            lock (this.registerSkippedLock)
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs:214]: This lock adds unnecessary overhead, the concurrent data structures are already thread-safe (and often times lock-free!)

File: src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs [hunk 3]
```diff
@@ -62,6 +62,24 @@ public IEnumerable<DetectedComponent> GetDetectedComponents()
...
 
+    public IEnumerable<string> GetSkippedComponents()
+    {
+        IEnumerable<string> skippedComponents;
+        if (this.singleFileRecorders == null)
+        {
+            return Enumerable.Empty<string>();
+        }
+
+        skippedComponents = this.singleFileRecorders
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs:73]: You can simpify this query: 

```csharp
this.singleFileRecorders.Select(x => x.GetSkippedComponents().Keys).SelectMany(x => x).Distinct().ToImmutableList();
```

File: src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs (hunk 1 above)
Comment (cobya) [reply src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs:120]: This is so we aren't iterating the entire list when searching for duplicates on list addition - this is fine for repos with a small number of components but can get expensive for large repositories.

File: src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs (hunk 3 above)
Comment (JamieMagee) [reply src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs:73]: You can also return directly here. No need to split declaration, initialization, and return statement.

**Comment by JamieMagee** [comment]
There are now some places where we log in the detector that we're not able to parse a package, and some places where we don't. We not also provide a summary of skipped or unparseable packages at the end of the run.

Is it necessary to log directly in the detector if we're providing a summary at the end?

File: src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs (hunk 1 above)
Comment (melotic) [reply src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs:120]: forgot .NET doesn't like concurrent hash sets :(

**Comment by cobya** [comment]
> There are now some places where we log in the detector that we're not able to parse a package, and some places where we don't. We not also provide a summary of skipped or unparseable packages at the end of the run.
> 
> Is it necessary to log directly in the detector if we're providing a summary at the end?

The purpose of having the summary at the end is not to provide detailed exception messages of why it has been skipped - that should be done in the detector. The summary is to quickly show users how many components were skipped so they can investigate further.

File: src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs (hunk 3 above)
Comment (cobya) [reply src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs]: Done.

File: src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs (hunk 2 above)
Comment (cobya) [reply src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs]: Done.

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**